### PR TITLE
fix #14574, `cp` on files >2GB

### LINKED ISSUE #14574: cp(file,file-copy) broken for >2GB files
Hello,

I'm not sure if this is an issue with my installation, but the `cp` function is not working for files larger than 2GB:

The following code:

``` julia
io = open("test.bin","w")
write(io,rand(Uint8,3_000_000_000)
close(io)

cp("test.bin","test-copy.bin", remove_destination=true)

@show filesize("test.bin")
@show filesize("test-copy.bin")
```

produces:

```
filesize("test.bin") = 3000000000
filesize("test-copy.bin") = 2147479552
```

I guess it might be a 32bit file pointer issue, so I thought I should also note that on my install, I see: `sizeof(Int)=8`.

Here's the rest of `versioninfo()`:

```
Julia Version 0.4.2
Commit bb73f34* (2015-12-06 21:47 UTC)
Platform Info:
  System: Linux (x86_64-redhat-linux)
  CPU: Intel(R) Core(TM) i5-3320M CPU @ 2.60GHz
  WORD_SIZE: 64
  BLAS: libopenblas (DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblasp.so.0
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```

Thanks!

Sam

### PR DESCRIPTION
**Pull request by JeffBezanson**
There may be a better way to do this, but this is a silly bug to have so here is a simple fix for now.

Also, any ideas on how to test this? Creating multiple GB of files seems like the kind of thing that could disagree with CI systems.

### CODE REVIEW

**Comment by ViralBShah** [comment]
Copy from /dev/zero to /dev/null?

File: base/file.jl [hunk 1]
```diff
@@ -746,8 +746,14 @@ function sendfile(src::AbstractString, dst::AbstractString)
         dst_file = open(dst, JL_O_CREAT | JL_O_TRUNC | JL_O_WRONLY, filemode(src_file))
         dst_open = true
 
-        bytes = filesize(stat(src_file))
-        sendfile(dst_file, src_file, Int64(0), Int(bytes))
+        bytes = Int(filesize(stat(src_file)))
+        offs = Int64(0)
```
Comment (vtjnash) [inline base/file.jl]: Set this to const -1

**Comment by JeffBezanson** [comment]
Ok, we can now use libuv's `copyfile`, but there's a problem. Previously we created the destination file with just `open`, which respects the process umask. But libuv `copyfile` copies the source file's mode exactly, independent of the umask. Should we change this behavior, or do an extra `chmod`?

**Comment by JeffreySarnoff** [comment]
When copying a file, keeping that file's mode seems more obvious (when I choose to edit a read-only file, it ought require some overtly permissive act).

**Comment by nalimilan** [comment]
FWIW Python 3 [preserves permissions](https://docs.python.org/3.7/library/shutil.html#shutil.copy). The fact that `uv_copy_file` also preserves them is probably intentional too.

Though for a backport we would probably need to preserve the previous behavior.

**Comment by JeffBezanson** [comment]
Good point, I'll do a backwards-compatible fix first.

**Comment by nalimilan** [comment]
Maybe we could have tests that are only run manually, before tagging a release? It sounds problematic to have code paths that are never exercized at all.

**Comment by ararslan** [comment]
So can this be backported?

**Comment by JeffBezanson** [comment]
Yes I think so.

**Comment by ararslan** [comment]
Looks like this didn't make it into 1.1.1. I'll apply the label for 1.2, then hopefully we can get it into rc2. I just hit this issue again today.

**Comment by JeffBezanson** [comment]
I believe this is already in 1.2.

**Comment by ararslan** [comment]
Hm, yeah, it does appear to be on the `release-1.2` branch. There don't appear to be any references to this PR though, so I guess backporting isn't including a PR reference as recommended by DISTRIBUTING.md?

**Comment by KristofferC** [comment]
What do you mean? This wasn't backported, it was merged when branching.

**Comment by ararslan** [comment]
Oh sorry, I'm just confused. Please ignore me.

**Pull request merged**